### ISSUE TITLE
Proposal out-of-tree cloud provider implementation/migration

### ISSUE THREAD

**Issue by mrIncompetent** [kubermatic/kubermatic#3345]
**User Story**
Since the in-tree cloud providers have been marked as deprecated, we should come up with a design/proposal on how we can integrate the out-of-tree cloud providers in Kubermatic.

Where do host them? User cluster or seed cluster?
How do we migrate from in-tree to out-of-tree?
What release cycle to the out-of-tree providers have?
What components do we need take care of? (CSI? CCM? More?)

Node:
The machine-controller supports already a flag to set `--cloud-provider=external` on the kubelet.

**Acceptance criteria**
- [ ] A design proposal has been written for the implementation & migration

**Comment by alvaroaleman** [comment]
> Where do host them? User cluster or seed cluster?

They must be in the seed, because nodes with `--cloud-provider=external` are marked as unschedulable until the CCM initialized them, so hosting it in the user cluster would pose a chicken-egg problem

> How do we migrate from in-tree to out-of-tree?

I don't think we can migrate existing clusters because we need to re-create all nodes. The only thing I can imagine is having a field in the `cluster` crd indicating weather the cluster uses in-tree/out-of-tree CCM and advising ppl to manually change that, then re-create all nodes

**Comment by alvaroaleman** [comment]
IMHO we should start doing that for one provider first (most likely Openstack, as its out of tree CCM is the most active one) and then go from there

**Comment by alvaroaleman** [comment]
Actually we can migrate clusters by changing their NodeDeployments to use the external cloud provider. This will likely require the `-external-cloud-provider` flag to be moved to a machineSpec property, because our migrations run before we start the kubermatic controller manager who would set that flag.

**Comment by alvaroaleman** [comment]
At the sig-openstack deep dive in Barcelona, the awnswer to "How to migrate from in-tree to out-of-tree providers" was "There is no migration path but an ongoing discussion how to solve this".

I propose to scope this issue down to: "How to integrate external cloud providers in a way that they will only be used for newly created clusters and old clusters continue to use the in-tree provider".

Concerns?

**Comment by mrIncompetent** [comment]
We can start with an cluster level feature flag & postpone a migration for now.
The proposal should mention that we need to postpone a full migration until upstream actually offers one.
